### PR TITLE
Adam does not need to copy grads to g_shared

### PR DESCRIPTION
**Pull request by jozef-mokry**
This might save some memory and copying time.

I tested the change by:
- training with adam and checking whether the parameters change over iterations
- training with adadelta and checking whether the parameters change over iterations

### CODE REVIEW

**Comment by rsennrich** [comment]
Hi Jozef,

thank you for this patch - it seems to work well, but my benchmark results are mixed.

running test_train.sh, I'm getting a reduction in peak memory consumption of 10%, but also a reduction of speed of the same magnitude:

baseline:
Epoch  21 Update  500 Cost  47.1492843628 UD  138.799925089 139.55 sentences/s

no_grads_shared:
Epoch  21 Update  500 Cost  47.182926178 UD  151.010597944 128.27 sentences/s

if this involves a speed-memory trade-off, I'm inclined to keep the faster version, maybe adding the low-memory version as a configurable option.

**Comment by jozef-mokry** [comment]
It's strange that it's slower. The only changes that I think could cause a slowdown are:
- cost is computed in the same function that also updates parameters and theano does a bad job at optimizing the function now
- there is an extra 'if' statement in the main training loop

I ran ./test_train.sh on GeForce GTX TITAN X and I got:
baseline:
Epoch  21 Update  500 Cost  47.1806678772 UD  122.921011925 157.58 sentences/s 
no_grads_shared:
Epoch  21 Update  500 Cost  47.178768157958984 UD  125.464458942 154.39 sentences/s

I had the following settings:
theano version: 0.9.0dev4.dev
cuda version: 8.0.44
cudnn version: 5.1
theano flags:
[global]
floatX = float32
device = gpu
mode=FAST_RUN
[lib]
cnmem = 1
[dnn]
enabled = True

When I don't use cnmem I get:
baseline:
Epoch  21 Update  500 Cost  47.1848258972 UD  162.454783916 119.23 sentences/s
(max memory 1339 MiB)
no_grads_shared:
Epoch  21 Update  500 Cost  47.1866455078 UD  165.711035967 116.89 sentences/s
(max memory 1068 MiB)

**Comment by jozef-mokry** [comment]
I am still getting a slight decrease in time performance so I agree we should probably leave it as it is.

**Comment by bastings** [comment]
Hi @jozef-mokry 
I think it would be too bad to close this PR already, since it really makes sense not to store the gradients in memory (there is no good reason to do it, except for educational purposes)

Couldn't it be that this line makes things slower?
``for p, g in zip(tparams.values(), grads):``

tparams.values() does not seem so efficient; we can use an iterator instead. same for grads?

**Comment by jozef-mokry** [comment]
Hi @bastings 
That could be the problem. 
However I think that this 'for' loop only runs when the optimizer is being built - i.e. it might make the compilation time longer, but I am not sure if it can affect the training speed.

**Comment by rsennrich** [comment]
The theano profiler shows some differences between the two branches (using GTX 1080):

master:
CUDA_LAUNCH_BLOCKING=1 THEANO_FLAGS=profile=True,mode=FAST_RUN,floatX=float32,device=gpu ./test_train.sh

Epoch  21 Update  500 Cost  47.1845092773 UD  165.825698137 116.81 sentences/s

<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  29.7%    29.7%      46.834s       9.37e-02s     Py     500        1   forall_inplace,gpu,grad_of_decoder__layers}
  14.8%    44.5%      23.367s       4.67e-02s     C      500        1   GpuDnnSoftmaxGrad{tensor_format='bc01', mode='channel', algo='accurate'}
   9.9%    54.4%      15.564s       3.11e-02s     Py     500        1   forall_inplace,gpu,decoder__layers}
   9.3%    63.7%      14.724s       1.13e-03s     C     13000       26   GpuCAReduce{pre=sqr,red=add}{1,1}
   6.9%    70.6%      10.852s       2.17e-02s     Py     500        1   forall_inplace,gpu,grad_of_encoder_r__layers}
   6.8%    77.3%      10.671s       2.13e-02s     Py     500        1   forall_inplace,gpu,grad_of_encoder__layers}
   5.4%    82.8%       8.567s       3.65e-04s     C     23500       47   GpuDot22
   2.4%    85.1%       3.733s       7.47e-03s     C      500        1   GpuAdvancedIncSubtensor1{inplace,inc}
   1.8%    87.0%       2.902s       5.80e-03s     Py     500        1   forall_inplace,gpu,encoder__layers}
   1.8%    88.8%       2.872s       5.74e-03s     Py     500        1   forall_inplace,gpu,encoder_r__layers}
   1.6%    90.4%       2.464s       1.20e-04s     C     20500       41   GpuElemwise{Composite{(i0 - ((i1 * i2) / (i3 + sqrt(i4))))}}[(0, 0)]
   1.2%    91.6%       1.962s       1.45e-04s     C     13500       27   GpuAlloc{memset_0=True}
   1.0%    92.6%       1.562s       3.12e-03s     C      500        1   GpuDnnSoftmax{tensor_format='bc01', mode='channel', algo='accurate'}
   0.9%    93.5%       1.442s       3.21e-04s     C     4500        9   GpuCAReduce{add}{1,1,0}
   0.7%    94.2%       1.155s       5.63e-05s     C     20500       41   GpuElemwise{Composite{((i0 * i1) + (i2 * i3))}}[(0, 1)]
   0.7%    95.0%       1.149s       1.44e-04s     C     8000       16   GpuElemwise{Composite{Switch(i0, (i1 / i2), i1)},no_inplace}
   0.7%    95.7%       1.123s       5.48e-05s     C     20500       41   GpuElemwise{Composite{((i0 * i1) + (i2 * sqr(i3)))}}[(0, 1)]
   0.7%    96.4%       1.067s       1.07e-03s     C     1000        2   GpuElemwise{Add}[(0, 0)]
   0.7%    97.0%       1.042s       2.89e-05s     C     36000       72   GpuReshape{2}
   0.5%    97.6%       0.853s       1.07e-04s     C     8000       16   GpuElemwise{mul,no_inplace}


 no_grads_shared:
CUDA_LAUNCH_BLOCKING=1 THEANO_FLAGS=profile=True,mode=FAST_RUN,floatX=float32,device=gpu ./test_train.sh

Epoch  21 Update  500 Cost  47.1848068237 UD  180.173623085 107.51 sentences/s

<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  28.3%    28.3%      47.502s       9.50e-02s     Py     500        1   forall_inplace,gpu,grad_of_decoder__layers}
  13.9%    42.2%      23.325s       4.66e-02s     C      500        1   GpuDnnSoftmaxGrad{tensor_format='bc01', mode='channel', algo='accurate'}
   9.5%    51.7%      16.023s       3.20e-02s     Py     500        1   forall_inplace,gpu,decoder__layers}
   8.8%    60.5%      14.707s       1.13e-03s     C     13000       26   GpuCAReduce{pre=sqr,red=add}{1,1}
   6.6%    67.1%      11.068s       2.21e-02s     Py     500        1   forall_inplace,gpu,grad_of_encoder_r__layers}
   6.5%    73.6%      10.958s       2.19e-02s     Py     500        1   forall_inplace,gpu,grad_of_encoder__layers}
   6.2%    79.8%      10.413s       4.43e-04s     C     23500       47   GpuDot22
   2.5%    82.2%       4.115s       8.23e-03s     C      500        1   GpuAdvancedIncSubtensor1{inplace,inc}
   2.0%    84.2%       3.361s       2.49e-04s     C     13500       27   GpuAlloc{memset_0=True}
   1.8%    86.0%       2.983s       5.97e-03s     Py     500        1   forall_inplace,gpu,encoder_r__layers}
   1.8%    87.8%       2.979s       5.96e-03s     Py     500        1   forall_inplace,gpu,encoder__layers}
   1.5%    89.3%       2.485s       1.21e-04s     C     20500       41   GpuElemwise{Composite{(i0 - ((i1 * i2) / (i3 + sqrt(i4))))}}[(0, 0)]
   1.3%    90.6%       2.231s       2.79e-04s     C     8000       16   GpuElemwise{mul,no_inplace}
   1.0%    91.6%       1.627s       4.52e-05s     C     36000       72   GpuReshape{2}
   0.9%    92.5%       1.506s       3.01e-03s     C      500        1   GpuDnnSoftmax{tensor_format='bc01', mode='channel', algo='accurate'}
   0.9%    93.3%       1.449s       3.22e-04s     C     4500        9   GpuCAReduce{add}{1,1,0}
   0.7%    94.0%       1.164s       1.45e-04s     C     8000       16   GpuElemwise{Composite{Switch(i0, (i1 / i2), i1)},no_inplace}
   0.7%    94.7%       1.159s       6.27e-05s     C     18500       37   GpuElemwise{Composite{((i0 * i1) + (i2 * i3))}}[(0, 1)]
   0.7%    95.4%       1.128s       2.82e-04s     C     4000        8   GpuElemwise{add,no_inplace}
   0.7%    96.0%       1.125s       6.08e-05s     C     18500       37   GpuElemwise{Composite{((i0 * i1) + (i2 * sqr(i3)))}}[(0, 1)]


not sure where does differences come from, and why I see bigger speed differences than Jozef between the branches.

**Comment by bastings** [comment]
Are you using the same Theano version?

**Comment by rsennrich** [comment]
The numbers I reported before were with Theano 0.8.2, but I also tested the dev version (0.9.0dev4.dev-7fd891), also on a GTX 1080. Using Theano 0.8 and cudnn-5.1 for all experiments.

master:
Epoch  21 Update  500 Cost  47.1829032898 UD  111.653115034 173.48 sentences/s
no_shared_grads:
Epoch  21 Update  500 Cost  47.1819877625 UD  127.213812828 152.26 sentences/s

**Comment by bastings** [comment]
These are the numbers for 0.9.0dev5 on Nvidia Geforce Titan X:

master:
Epoch  21 Update  500 Cost  47.1810379028 UD  122.972671032 157.51 sentences/s

dont_copy_grads
Epoch  21 Update  500 Cost  47.185218811 UD  124.091377974 156.09 sentences/s

So also a small difference. I might try a few changes to the branch to see if I can make it faster.

**Comment by bastings** [comment]
By the way -- i'm using CUDA8 on Ubuntu. Did both of you use that too? 

So the summary is that for @jozef-mokry and me the numbers are really acceptable, but for you @rsennrich the different is too big. The problem seems really setup related to me. I checked the code carefully but I do not see clear ways to improve (training) speed.

**Comment by rsennrich** [comment]
OK, the speed difference is confirmed to be smaller on 0.9.0dev5 than on 0.8.2 (using the new GPU backend):

master:
Epoch  21 Update  500 Cost  47.1839561462 UD  93.0547020435 208.16 sentences/s

dont_copy_grads:
Epoch  21 Update  500 Cost  47.1829147339 UD  94.4189739227 205.15 sentences/s

thanks for looking into possibly speeding up dont_copy_grads. If that's not possible, I'd suggest pulling this in once theano 0.9 is released.

**Comment by bastings** [comment]
Yes, let's wait for Theano 0.9 and then pull

**Comment by rsennrich** [comment]
theano 0.9 is now officially out, so I've merged this. I encourage people to upgrade to Theano 0.9.

**Comment by Koeldc** [comment]
File "build/bdist.linux-x86_64/egg/nematus/nmt.py", line 1147, in train
  File "build/bdist.linux-x86_64/egg/nematus/util.py", line 18, in load_dict

Do I change the train.sh file like this?

**device=cuda

THEANO_FLAGS='floatX=float32,device=$device,gpurray.preallocate=1' python config.py
**

**Comment by rsennrich** [comment]
yes, that looks fine. You can see here which options we used for training systems at WMT17: http://data.statmt.org/wmt17_systems/training/
